File: gaming_deals_bot/formatter.py
"""Message formatting for Matrix deal posts.

Produces both a plain-text fallback and HTML formatted body for Matrix messages.
HTML is built directly using the Matrix-supported HTML subset rather than going
through a Markdown-to-HTML converter, which collapses line breaks.
"""

from datetime import datetime, timezone
from html import escape

from .cheapshark import CheapSharkDeal
from .currency import format_price
from .epic import EpicFreeGame
from .itad_deals import ITADDeal


def format_deal(deal: CheapSharkDeal, is_historical_low: bool = False) -> tuple[str, str]:
    """Format a CheapShark deal into (plain_text, html) for Matrix.

    Returns (body, formatted_body).
    """
    discount = int(deal.savings)
    sale_price = float(deal.sale_price)
    normal_price = float(deal.normal_price)

    sale_multi = format_price(sale_price)
    normal_display = format_price(normal_price)
    title = escape(deal.title)

    html_lines = [
        f"<strong>🎮 [DEAL] {title}</strong>",
        f"{discount}% off on {escape(deal.store_name)} <del>{escape(normal_display)}</del>",
        f"💰 <strong>{escape(sale_multi)}</strong>",
    ]
    if is_historical_low:
        html_lines.append("🏆 <em>All-time low!</em>")
    html_lines.append(f'🔗 <a href="{escape(deal.deal_url)}">View Deal</a>')
    html = "<br>\n".join(html_lines)

    plain_lines = [
        f"🎮 [DEAL] {deal.title}",
        f"  {discount}% off on {deal.store_name} (was {normal_display})",
        f"  💰 {sale_multi}",
    ]
    if is_historical_low:
        plain_lines.append("  🏆 All-time low!")
    plain_lines.append(f"  🔗 {deal.deal_url}")
    plain_text = "\n".join(plain_lines)

    return plain_text, html


def format_itad_deal(deal: ITADDeal) -> tuple[str, str]:
    """Format an ITAD deal into (plain_text, html) for Matrix.

    Returns (body, formatted_body).
    """
    sale_multi = format_price(deal.sale_price)
    normal_display = format_price(deal.normal_price)
    title = escape(deal.title)

    html_lines = [
        f"<strong>🎮 [DEAL] {title}</strong>",
        f"{deal.discount}% off on {escape(deal.shop_name)} <del>{escape(normal_display)}</del>",
        f"💰 <strong>{escape(sale_multi)}</strong>",
    ]
    if deal.is_historical_low:
        html_lines.append("🏆 <em>All-time low!</em>")
    html_lines.append(f'🔗 <a href="{escape(deal.url)}">View Deal</a>')
    html = "<br>\n".join(html_lines)

    plain_lines = [
        f"🎮 [DEAL] {deal.title}",
        f"  {deal.discount}% off on {deal.shop_name} (was {normal_display})",
        f"  💰 {sale_multi}",
    ]
    if deal.is_historical_low:
        plain_lines.append("  🏆 All-time low!")
    plain_lines.append(f"  🔗 {deal.url}")
    plain_text = "\n".join(plain_lines)

    return plain_text, html


def format_epic_free(game: EpicFreeGame) -> tuple[str, str]:
    """Format an Epic free game into (plain_text, html) for Matrix.

    Returns (body, formatted_body).
    """
    title = escape(game.title)

    html_lines = [
        f"<strong>🆓 [FREE] {title}</strong>",
        "Free on Epic Games Store",
    ]

    if game.end_date:
        try:
            end_dt = datetime.fromisoformat(game.end_date.replace("Z", "+00:00"))
            date_str = end_dt.strftime("%B %-d")
            html_lines.append(f"📅 <em>Free until {date_str}</em>")
        except (ValueError, TypeError):
            pass

    if game.url:
        html_lines.append(f'🔗 <a href="{escape(game.url)}">Claim Now</a>')
    html = "<br>\n".join(html_lines)

    plain_lines = [
        f"🆓 [FREE] {game.title}",
        "  Free on Epic Games Store",
    ]
    if game.end_date:
        try:
            end_dt = datetime.fromisoformat(game.end_date.replace("Z", "+00:00"))
            date_str = end_dt.strftime("%B %-d")
            plain_lines.append(f"  📅 Free until {date_str}")
        except (ValueError, TypeError):
            pass
    if game.url:
        plain_lines.append(f"  🔗 {game.url}")
    plain_text = "\n".join(plain_lines)

    return plain_text, html


def format_epic_upcoming(game: EpicFreeGame) -> tuple[str, str]:
    """Format an upcoming Epic free game into (plain_text, html) for Matrix."""
    title = escape(game.title)

    html_lines = [
        f"<strong>📢 [UPCOMING FREE] {title}</strong>",
        "Coming soon — Free on Epic Games Store",
    ]

    if game.end_date:
        try:
            end_dt = datetime.fromisoformat(game.end_date.replace("Z", "+00:00"))
            date_str = end_dt.strftime("%B %-d")
            html_lines.append(f"📅 <em>Free until {date_str}</em>")
        except (ValueError, TypeError):
            pass

    if game.url:
        html_lines.append(f'🔗 <a href="{escape(game.url)}">Store Page</a>')
    html = "<br>\n".join(html_lines)

    plain_lines = [
        f"📢 [UPCOMING FREE] {game.title}",
        "  Coming soon — Free on Epic Games Store",
    ]
    if game.url:
        plain_lines.append(f"  🔗 {game.url}")
    plain_text = "\n".join(plain_lines)

    return plain_text, html

File: internal/formatter/formatter.go
package formatter

import (
	"fmt"
	"html"
	"math"
	"strings"

	"github.com/prosolis/Pastel/internal/currency"
	"github.com/prosolis/Pastel/internal/deals"
)

// Message holds both plain-text and HTML versions of a formatted message.
type Message struct {
	Plain string
	HTML  string
}

// FormatCheapSharkDeal formats a CheapShark deal for Matrix.
func FormatCheapSharkDeal(d deals.CheapSharkDeal, conv *currency.Converter) Message {
	discount := int(math.Floor(d.Savings))
	saleMulti := conv.FormatPrice(d.SalePrice)
	normalMulti := conv.FormatPrice(d.NormalPrice)

	var plain, htmlB strings.Builder

	plain.WriteString(fmt.Sprintf("🎮 [DEAL] %s\n", d.Title))
	plain.WriteString(fmt.Sprintf("  %d%% off on %s (was %s)\n", discount, d.StoreName, normalMulti))
	plain.WriteString(fmt.Sprintf("  💰 %s\n", saleMulti))

	htmlB.WriteString(fmt.Sprintf("<strong>🎮 [DEAL] %s</strong><br>\n", html.EscapeString(d.Title)))
	htmlB.WriteString(fmt.Sprintf("%d%% off on %s <del>%s</del><br>\n",
		discount, html.EscapeString(d.StoreName), html.EscapeString(normalMulti)))
	htmlB.WriteString(fmt.Sprintf("💰 <strong>%s</strong><br>\n", html.EscapeString(saleMulti)))

	if d.IsHistLow {
		plain.WriteString("  🏆 All-time low!\n")
		htmlB.WriteString("🏆 <em>All-time low!</em><br>\n")
	}

	plain.WriteString(fmt.Sprintf("  🔗 %s", d.DealURL))
	htmlB.WriteString(fmt.Sprintf("🔗 <a href=\"%s\">View Deal</a>", html.EscapeString(d.DealURL)))

	return Message{Plain: plain.String(), HTML: htmlB.String()}
}

// FormatITADDeal formats an ITAD deal for Matrix.
func FormatITADDeal(d deals.ITADDeal, conv *currency.Converter) Message {
	saleMulti := conv.FormatPrice(d.Price)
	normalMulti := conv.FormatPrice(d.Regular)

	var plain, htmlB strings.Builder

	plain.WriteString(fmt.Sprintf("🎮 [DEAL] %s\n", d.Title))
	plain.WriteString(fmt.Sprintf("  %d%% off on %s (was %s)\n", d.Discount, d.ShopName, normalMulti))
	plain.WriteString(fmt.Sprintf("  💰 %s\n", saleMulti))

	htmlB.WriteString(fmt.Sprintf("<strong>🎮 [DEAL] %s</strong><br>\n", html.EscapeString(d.Title)))
	htmlB.WriteString(fmt.Sprintf("%d%% off on %s <del>%s</del><br>\n",
		d.Discount, html.EscapeString(d.ShopName), html.EscapeString(normalMulti)))
	htmlB.WriteString(fmt.Sprintf("💰 <strong>%s</strong><br>\n", html.EscapeString(saleMulti)))

	if d.IsHistLow {
		plain.WriteString("  🏆 All-time low!\n")
		htmlB.WriteString("🏆 <em>All-time low!</em><br>\n")
	}

	plain.WriteString(fmt.Sprintf("  🔗 %s", d.URL))
	htmlB.WriteString(fmt.Sprintf("🔗 <a href=\"%s\">View Deal</a>", html.EscapeString(d.URL)))

	return Message{Plain: plain.String(), HTML: htmlB.String()}
}

// FormatEpicFreeGame formats an Epic free game for Matrix.
func FormatEpicFreeGame(g deals.EpicFreeGame) Message {
	var plain, htmlB strings.Builder

	if g.Upcoming {
		plain.WriteString(fmt.Sprintf("📢 [UPCOMING FREE] %s\n", g.Title))
		plain.WriteString("  Coming soon — Free on Epic Games Store\n")

		htmlB.WriteString(fmt.Sprintf("<strong>📢 [UPCOMING FREE] %s</strong><br>\n", html.EscapeString(g.Title)))
		htmlB.WriteString("Coming soon — Free on Epic Games Store<br>\n")
	} else {
		plain.WriteString(fmt.Sprintf("🆓 [FREE] %s\n", g.Title))
		plain.WriteString("  Free on Epic Games Store\n")

		htmlB.WriteString(fmt.Sprintf("<strong>🆓 [FREE] %s</strong><br>\n", html.EscapeString(g.Title)))
		htmlB.WriteString("Free on Epic Games Store<br>\n")
	}

	if g.EndDate != nil {
		dateStr := g.EndDate.Format("January 2")
		plain.WriteString(fmt.Sprintf("  📅 Free until %s\n", dateStr))
		htmlB.WriteString(fmt.Sprintf("📅 <em>Free until %s</em><br>\n", html.EscapeString(dateStr)))
	}

	linkText := "Claim Now"
	if g.Upcoming {
		linkText = "Store Page"
	}

	plain.WriteString(fmt.Sprintf("  🔗 %s", g.URL))
	htmlB.WriteString(fmt.Sprintf("🔗 <a href=\"%s\">%s</a>", html.EscapeString(g.URL), linkText))

	return Message{Plain: plain.String(), HTML: htmlB.String()}
}

// FormatWatchlistNotification formats a DM notification for a watchlist match.
func FormatWatchlistNotification(watchedName, dealTitle, dealURL string, price string, discount int) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("🔔 Deal alert for \"%s\"!\n", watchedName))
	sb.WriteString(fmt.Sprintf("  %s — %d%% off\n", dealTitle, discount))
	sb.WriteString(fmt.Sprintf("  💰 %s\n", price))
	sb.WriteString(fmt.Sprintf("  🔗 %s", dealURL))
	return sb.String()
}

// FormatWatchlistFreeNotification formats a DM notification for a free game watchlist match.
func FormatWatchlistFreeNotification(watchedName, dealTitle, dealURL string) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("🔔 Free game alert for \"%s\"!\n", watchedName))
	sb.WriteString(fmt.Sprintf("  %s — Free on Epic Games Store\n", dealTitle))
	sb.WriteString(fmt.Sprintf("  🔗 %s", dealURL))
	return sb.String()
}
